pkg/storage/etcd3/store.go

```diff
@@ -0,0 +1,410 @@
+	"golang.org/x/net/context"
+)
+
+type store struct {
+	client     *clientv3.Client
+	codec      runtime.Codec
+	versioner  storage.Versioner
+	pathPrefix string
+}
+
+type elemForDecode struct {
+	data []byte
+	rev  uint64
+}
+
+type objState struct {
+	obj  runtime.Object
+	meta *storage.ResponseMeta
+	rev  int64
+	data []byte
+}
+
+func newStore(c *clientv3.Client, codec runtime.Codec, prefix string) *store {
+	return &store{
+		client:     c,
+		versioner:  etcd.APIObjectVersioner{},
+		codec:      codec,
+		pathPrefix: prefix,
+	}
+}
+
+func (s *store) Backends(ctx context.Context) []string {
+	resp, err := s.client.MemberList(ctx)
+	if err != nil {
+		glog.Errorf("Error obtaining etcd members list: %q", err)
+		return nil
+	}
+	var mlist []string
+	for _, member := range resp.Members {
+		mlist = append(mlist, member.ClientURLs...)
+	}
+	return mlist
+}
+
+func (s *store) Codec() runtime.Codec {
+	return s.codec
+}
+
+func (s *store) Versioner() storage.Versioner {
+	return s.versioner
+}
+
+func (s *store) Get(ctx context.Context, key string, out runtime.Object, ignoreNotFound bool) error {
+	key = keyWithPrefix(s.pathPrefix, key)
+	getResp, err := s.client.KV.Get(ctx, key)
+	if err != nil {
+		return err
+	}
+
+	if len(getResp.Kvs) == 0 {
+		if ignoreNotFound {
+			return runtime.SetZeroValue(out)
+		}
+		return storage.NewKeyNotFoundError(key, 0)
+	}
+	kv := getResp.Kvs[0]
+	return decode(s.codec, s.versioner, kv.Value, out, kv.ModRevision)
+}
+
+func (s *store) Create(ctx context.Context, key string, obj, out runtime.Object, ttl uint64) error {
+	if version, err := s.versioner.ObjectResourceVersion(obj); err == nil && version != 0 {
+		return errors.New("resourceVersion should not be set on objects to be created")
+	}
+	data, err := runtime.Encode(s.codec, obj)
+	if err != nil {
+		return err
+	}
+	key = keyWithPrefix(s.pathPrefix, key)
+
+	txnResp, err := s.client.KV.Txn(ctx).If(
+		notFound(key),
+	).Then(
+		clientv3.OpPut(key, string(data)),
+	).Commit()
+	if err != nil {
+		return err
+	}
+	if !txnResp.Succeeded {
+		return storage.NewKeyExistsError(key, 0)
+	}
+
+	if out != nil {
+		putResp := txnResp.Responses[0].GetResponsePut()
+		return decode(s.codec, s.versioner, data, out, putResp.Header.Revision)
+	}
+	return nil
+}
+
+func (s *store) Delete(ctx context.Context, key string, out runtime.Object, precondtions *storage.Preconditions) error {
+	v, err := conversion.EnforcePtr(out)
+	if err != nil {
+		panic("unable to convert output object to pointer")
+	}
+	key = keyWithPrefix(s.pathPrefix, key)
+	if precondtions == nil {
+		return s.unconditionalDelete(ctx, key, out)
+	}
+	return s.conditionalDelete(ctx, key, out, v, precondtions)
+}
+
+func (s *store) unconditionalDelete(ctx context.Context, key string, out runtime.Object) error {
+	// We need to do get and delete in single transaction in order to
+	// know the value and revision before deleting it.
+	txnResp, err := s.client.KV.Txn(ctx).If().Then(
+		clientv3.OpGet(key),
```

Comment:
@timothysc Thanks! That would be great. Previously, it is impossible for etcd to get previous after a deletion or modification. Now etcd supports multiple versioned key, so it is possible for the application to get the previous value when they want. We will probably add a shortcut into etcd to save a RTT, but it will be an optional optimization.